torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler/transpose.h>
+
+#include <torch/csrc/jit/codegen/cuda/executor_utils.h>
+#include <torch/csrc/jit/codegen/cuda/inline_propagator.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/pointwise_utils.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/registry.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/utils.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h>
+#include <torch/csrc/jit/codegen/cuda/transform_replay.h>
+#include <torch/csrc/jit/codegen/cuda/utils.h>
+
+#include <ATen/cuda/CUDAContext.h>
+
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+constexpr int64_t kThreadsPerBlock = 128;
+
+// DomainMap uses the ComputeAtMap to find a reference TensorView
+// that maps to all iterDomains in the fusion.
+class DomainMap : public pointwise_utils::DomainMap {
+ public:
+  using pointwise_utils::DomainMap::DomainMap;
+
+  TensorView* findReferenceFor(const std::vector<TensorView*>& group) const {
+    TensorView* result;
+    int max_dims = -1;
+    for (auto tv : group) {
+      if (isValidReference(tv)) {
+        int dims = pointwise_utils::nRootDims(tv);
+        if (dims > max_dims) {
+          result = tv;
+          max_dims = dims;
+        }
+      }
+    }
+    return result;
+  }
+
+  static bool hasAtLeastTwoValidGroups(Fusion* fusion) {
+    FusionGuard fg(fusion);
+    DomainMap domain_map(fusion);
+    auto grouped_inputs_outputs = domain_map.groupInputsOutputsByInnerDim();
+    if (grouped_inputs_outputs.size() < 2) {
+      return false;
+    }
+    return domain_map.findReferenceFor(grouped_inputs_outputs[0]) != nullptr &&
+        domain_map.findReferenceFor(grouped_inputs_outputs[1]) != nullptr;
+  }
+
+  int getPosMappedTo(TensorView* tv, IterDomain* id) {
+    const auto& dom = tv->domain()->domain();
+    for (auto i : c10::irange(dom.size())) {
+      if (areMapped(tv->axis(i), id)) {
+        return i;
+      }
+    }
+    TORCH_INTERNAL_ASSERT(
+        false, "Can not find ID mapped to ", id, " in tensor ", tv);
+  }
+
+  // Group inputs and outputs of a fusion by its inner most domain. For example
+  //   inputs: t0, t1
+  //   t2 = transpose(t1)
+  //   t3 = t0 + t2
+  //   t4 = sin(t0)
+  //   t5 = cos(t1)
+  //   outputs: t3, t4, t5
+  //
+  // Then we should have group {t0, t3, t4} and {t1, t5}
```

Comment:
I think this is a feature rather than a bug. For a fusion like `FusionScheduleBroadcastOnly_CUDA` the transpose scheduler provides better performance than the pointwise scheduler.

Reply:
I think for this case, it will create groups `{tv0}, {tv1}, {tv2}`, and two of them will be selected, and the other will be ignored. Currently, I am sorting groups based on number of tensors in the group, but looks like a better approach is based on "number of bytes". Will merge this PR as is, and resolve this in future PR.

Reply:
Sure, I was just thinking we could consider de-emphasize tensors that we can effectively treat as broadcast.